kubeadm: Make etcd member removal idempotent

dlipovetsky opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:
With this change, if the etcd member is not found, then it has already been removed, and `kubeadm reset` immediately completes the `remove-etcd-member` phase.

Previously, the phase would complete only once the exponential-backoff retry expired, up to 3 minutes duration. Below is what this looks like; note that `kubeadm reset` fails to delete the member ID `0`, but eventually completes the `remove-etcd-member` phase, and then the remaining phases:

```console
> kubeadm reset -f -v 5
I0425 07:42:54.260045   17919 reset.go:100] [reset] Loaded client set from kubeconfig file: /etc/kubernetes/admin.conf
[reset] Reading configuration from the cluster...
[reset] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -o yaml'
I0425 07:42:54.286968   17919 kubelet.go:74] attempting to download the KubeletConfiguration from ConfigMap "kubelet-config"
[preflight] Running pre-flight checks
I0425 07:42:54.303838   17919 reset.go:124] [reset] Detected and using CRI socket: unix:///run/containerd/containerd.sock
I0425 07:42:54.303875   17919 removeetcdmember.go:56] [reset] Checking for etcd config
I0425 07:42:54.303892   17919 local.go:93] [etcd] creating etcd client that connects to etcd pods
I0425 07:42:54.303905   17919 etcd.go:168] retrieving etcd endpoints from "kubeadm.kubernetes.io/etcd.advertise-client-urls" annotation in etcd Pods
I0425 07:42:54.308865   17919 etcd.go:104] etcd endpoints read from pods: https://10.0.130.65:2379,https://10.0.131.205:2379,https://10.0.133.32:2379
I0425 07:42:54.319603   17919 etcd.go:224] etcd endpoints read from etcd: https://10.0.131.205:2379,https://10.0.133.32:2379
I0425 07:42:54.319618   17919 etcd.go:122] update etcd endpoints: https://10.0.131.205:2379,https://10.0.133.32:2379
I0425 07:42:54.325820   17919 local.go:118] [etcd] get the member id from peer: https://10.0.130.65:2380
I0425 07:42:54.333868   17919 local.go:124] [etcd] removing etcd member: https://10.0.130.65:2380, id: 0
{"level":"warn","ts":"2023-04-25T07:42:54.342Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0005b56c0/10.0.131.205:2379","attempt":0,"error":"rpc error: code = NotFound desc = etcdserver: member not found"}
...
I0425 07:46:18.670046   17919 etcd.go:327] Failed to remove etcd member: etcdserver: member not found
W0425 07:46:18.670088   17919 removeetcdmember.go:64] [reset] Failed to remove etcd member: etcdserver: member not found, please manually remove this etcd member using etcdctl
I0425 07:46:18.670104   17919 cleanupnode.go:61] [reset] Getting init system
[reset] Stopping the kubelet service
```

This change also fixes a semantic bug in `etcd.GetMemberID`. The function now returns an error if no member ID is found for the peer. Previously, the function returned 0 if no member was found, but 0 is not a valid member ID.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:
This change does not change the existing behavior of the `remove-etcd-member` phase. It only shortens the amount of time that `kubeadm reset` spends in that phase when the etcd member ID is not found for the peer.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
If `kubeadm reset` finds no etcd member ID for the peer it removes during the `remove-etcd-member` phase, it continues immediately to other phases, instead of retrying the phase for up to 3 minutes before continuing.  
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dlipovetsky committed:
kubeadm: Make etcd member removal idempotent

If the etcd member is not found, then it has already been removed, and
kubeadm reset should immediately complete the 'remove-etcd-member'
phase. Previously, the phase would complete only once the
exponential-backoff retry expired, up to 3 minutes duration.

This commit also fixes a semantic error in etcd.GetMemberID. Previously,
the function returned 0 if no member was found, but 0 is not a valid
member ID.

dlipovetsky commented:
/sig cluster-lifecycle

dlipovetsky commented:
/cc @neolit123

dlipovetsky commented:
I mistakenly had a double import (one already existed under an alias). Because it's a small change, I'll just amend my commit, and force-push.

neolit123 commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -283,7 +285,7 @@ func (c *Client) GetMemberID(peerURL string) (uint64, error) {
 		}
 	}
-	return 0, nil
+	return 0, ErrNoMemberIDForPeerURL
```
this seems valid, but i wonder if it will break something. are we running GetMemberID() in more places where we expect no error? if yes, we should probably adapt all code around it.

dlipovetsky replied:
Thanks for your review! This is a great question.

The only other call site is https://github.com/kubernetes/kubernetes/blob/65e306d7d85dcfcbb9e3f5e7373f5469cc4437c7/cmd/kubeadm/app/phases/etcd/local.go#L180.

I've mentally walked this code.

Previous to this change, if `GetMemberID` did not find a member ID for the peer, it returned `0`. Afterward, `MemberPromote` logged the message `[etcd] Member 0 already promoted.` The function continued to run, possibly returning an error later from the call to `WaitForClusterAvailable`.

With this change, if `GetMemberID` does not find the member ID, it returns an error, and that error is immediately returned.

neolit123 replied:
ack

pacoxu replied:
This looks valid.

Only the learner mode will use it.
https://github.com/kubernetes/kubernetes/blob/28247d53d22bbea9344305ac3db53c473c6d6278/cmd/kubeadm/app/phases/etcd/local.go#L175-L184

The change is OK if the GetMemberID returns the error, it is not valid and should return err. Or the promote will fail bellow.

neolit123 reviewed (commented):
/approve
/hold

thanks, @dlipovetsky the change seems ok, but needs more eyes because this is a sensitive area and we don't have sufficient tests for all cases.

/cc @SataQiu @pacoxu

dlipovetsky commented:
> we don't have sufficient tests for all cases.

I noticed that we don't have tests that exercise the etcd client.

I think we could address this with an integration test, i.e. run 3 etcd processes. I _think_ that should work in prow...

neolit123 commented:
> > we don't have sufficient tests for all cases.
> 
> I noticed that we don't have tests that exercise the etcd client.
> 
> I think we could address this with an integration test, i.e. run 3 etcd processes. I _think_ that should work in prow...

do you mean the etcd client in etcd? we only use kinder for testing kubeadm binary calls, with the exception of some CLI integration tests in k/k. basing tests on kinder is a req since we'd ideally want to keep all test jobs generated and in the same place. not sure how we can tell the etcd client in kubeadm to do any arbitrary action foo.

IMO ideally we should have some kubeadm etcd code unit tests with fake etcd server / client. not sure to what extent etcd code supports that.

pacoxu commented on cmd/kubeadm/app/phases/etcd/local.go:
```diff
@@ -118,6 +118,10 @@ func RemoveStackedEtcdMemberFromCluster(client clientset.Interface, cfg *kubeadm
+		if errors.Is(etcdutil.ErrNoMemberIDForPeerURL, err) {
+			klog.V(5).Infof("[etcd] member was already removed, because no member id exists for peer %s", etcdPeerAddress)
+			return nil
+		}
```
The `RemoveMember` will handle `ErrNoMemberIDForPeerURL` later in the function.

Can we pass through so that the log of removing an etcd member will be printed and for file delete, not-found is a successful state? I don't have a strong opinion on this, but I believe it would make the log more readable for users.

pacoxu replied:
This change is inside `func RemoveStackedEtcdMemberFromCluster`.

With the current change, the log (if verbose is less than 5) will show only `[etcd] get the member id from peer: %s` and return.
I prefer to change the log level of line 122 higher so that users can know why.

Sorry. My last comment was wrong. If there is an error here, the remove member will use an empty ID which is odd.

dlipovetsky replied:
Thanks for rephrasing!

> I prefer to change the log level of line 122 higher so that users can know why.

I chose log level 5 for two reasons. First, because it's not an error, so I don't want to alert the user unless they've asked for increased verbosity. Second, because all log messages that indicate a problem (i.e. start with `Failed`) use level 5.

Of course I can reduce the log level, if you feel strongly. Let me know.

pacoxu replied:
/lgtm
overall

dlipovetsky commented:
@neolit123 I created a fake etcd client and wrote a unit test of `GetMemberID`: https://github.com/kubernetes/kubernetes/compare/master...dlipovetsky:kubernetes:kubeadm-etcd-client-refactor.

I _could_ add this to my PR, but I'm not sure I should. WDYT?

neolit123 commented:
new PR for these tests seems better

pacoxu reviewed (approved):
/priority important-longterm
/triage accepted

Only a log level need to be confirmed in https://github.com/kubernetes/kubernetes/pull/117724#discussion_r1185783073.

dlipovetsky commented:
I noticed that the log message I added did not write the member ID using hexadecimal format. Other log messages do. I fixed the formatting in a fixup commit.

SataQiu commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -326,6 +328,10 @@ func (c *Client) RemoveMember(id uint64) ([]Member, error) {
 			return true, nil
 		}
+		if errors.Is(rpctypes.ErrMemberNotFound, err) {
+			klog.V(5).Infof("Member was already removed, because member %016x was not found", id)
```
```suggestion
			klog.V(5).Infof("Member was already removed, because member %s was not found", strconv.FormatUint(id, 16))
```
Member id does not necessarily seems to be 16-bit.  We can use `strconv.FormatUint(id, 16)` here to format it, just like what `etcd` is doing:
https://github.com/kubernetes/kubernetes/blob/8f3f997a98c7b44925f7654e5c66e6f46e328147/vendor/go.etcd.io/etcd/client/pkg/v3/types/id.go#L24C7-L26

dlipovetsky replied:
I see what you mean. `strconv.FormatUint(id, 16)` does not pad the output with zeros, whereas `%016x` does, if there are fewer than 16 bytes. (If anyone is curious, see https://go.dev/play/p/pwOK24gL4nH)

In this file, there are 5 other log messages that use `%016x`, and I want to be consistent with them.

While I agree that we want use the upstream convention (no zero padding), I think it's a change that should be made to all log messages, and so is better as a separate PR.

dlipovetsky replied:
Also, thanks for the review @SataQiu!

SataQiu commented:
/approve

dlipovetsky commented:
I squashed the fixup.

dlipovetsky commented:
I rebased on master to get the unit test added in https://github.com/kubernetes/kubernetes/pull/117792.

Reviewers: The code you reviewed is in the first commit, and the rebase did not change it.

I added a fixup that updates the unit test.

I'll squash the fixup once I get an /lgtm.

dlipovetsky commented:
@neolit123 I think this has gotten reviews you asked for, and it includes new unit tests from https://github.com/kubernetes/kubernetes/pull/117792. Are you ok to cancel the /hold?

neolit123 commented:
/hold cancel
